messages/api.go

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) 2019 NEC Laboratories Europe GmbH.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package messages defines interface for the protocol messages.
+//
+// Note that some interfaces are "sealed" with unexported methods.
+// This ensures the desired semantic of type assertions/switches with
+// those interfaces.
+//
+// To fulfill a sealed interface in another package, the new type
+// needs to explicitly embed a dedicated type that implements the
+// interface's unexported methods. For example, consider the following
+// definition of a sealed interface:
+//
+// 	package api
+//
+//	type Sealed interface {
+// 		DoSomething()
+// 		isSealed()
+// 	}
+//
+//	type IsSealed struct{}
+//
+//	func (IsSealed) isSealed() {}
+//
+// A proper way to fulfill the Sealed interface in another package is
+// to define the new type like follows:
+//
+//	package mysealed
+//
+// 	type MySealed struct {
+// 		api.IsSealed
+// 		// ...
+// 	}
+//
+//	func (*MySealed) DoSomething() { /* ... */ }
+//
+package messages
+
+import (
+	"encoding"
+)
+
+// MessageImpl provides an implementation of the message representation.
+type MessageImpl interface {
+	NewFromBinary(data []byte) (Message, error)
+	NewRequest(clientID uint32, sequence uint64, operation []byte) Request
+	NewPrepare(replicaID uint32, view uint64, request Request) Prepare
+	NewCommit(replicaID uint32, prepare Prepare) Commit
+	NewReply(replicaID, clientID uint32, sequence uint64, result []byte) Reply
+}
+
+type Message interface {
+	encoding.BinaryMarshaler
+}
```

Comment:
This works for decoupling logic and representation, so I'm looking forward to see the concrete type.